lectures/opt_invest.md

```diff
@@ -0,0 +1,610 @@
+
+    # Compute and return v[i, j] - β Σ_jp v[σ[i, j], jp] * Q[j, jp]
+    return v - β * jnp.sum(V * Q, axis=2)
+
+R_σ = jax.jit(R_σ, static_argnums=(3,))
+```
+
++++ {"user_expressions": []}
+
+Define a function to get the value $v_{\sigma}$ of policy
+$\sigma$ by inverting the linear map $R_{\sigma}$.
+
+```{code-cell} ipython3
+def get_value(σ, constants, sizes, arrays):
+
+    # Unpack
+    β, a_0, a_1, γ, c = constants
+    y_size, z_size = sizes
+    y_grid, z_grid, Q = arrays
+
+    r_σ = compute_r_σ(σ, constants, sizes, arrays)
+
+    # Reduce R_σ to a function in v
+    partial_R_σ = lambda v: R_σ(v, σ, constants, sizes, arrays)
+
+    return jax.scipy.sparse.linalg.bicgstab(partial_R_σ, r_σ)[0]
+
+get_value = jax.jit(get_value, static_argnums=(2,))
+```
+
++++ {"user_expressions": []}
+
+Now, we compute the transition probabilities across states as a multi-index array
+
+$$
+         P_\sigma(y, z, y', z') = 1\{y' = \sigma(y, z)\} Q(z, z')
+$$
+
+```{code-cell} ipython3
+def compute_P_σ(σ, constants, sizes, arrays):
+
+    # Unpack model
+    β, a_0, a_1, γ, c = constants
+    y_size, z_size = sizes
+    y_grid, z_grid, Q = arrays
+
+    yp_idx = jnp.arange(y_size)
+    yp_idx = jnp.reshape(yp_idx, (1, 1, y_size, 1))
+    σ = jnp.reshape(σ, (y_size, z_size, 1, 1))
+    A = jnp.where(σ == yp_idx, 1, 0)
+    Q = jnp.reshape(Q, (1, z_size, 1, z_size))
+    P_σ = A * Q
+    return P_σ
+
+compute_P_σ = jax.jit(compute_P_σ, static_argnums=(2,))
+```
+
++++ {"user_expressions": []}
+
+Let's write a matrix version to compute the value function.
+
+We get the value v_σ of policy σ via
+
+$$
+        v_{\sigma} = (I - \beta P_{\sigma})^{-1} r_{\sigma}
+$$
+
+In this version we flatten the multi-index $[i, j]$ for the state $(y, z)$ to
+a single index $m$ and compute the vector $r_{\sigma}[m]$ and matrix $P_{\sigma}[m, mp]$
+giving transition probabilities across the single-index state.  Then we
+solve the above equation using matrix inversion.
+
+```{code-cell} ipython3
+def get_value_matrix_version(σ, constants, sizes, arrays):
+
+    # Unpack
+    β, a_0, a_1, γ, c = constants
+    y_size, z_size = sizes
+    y_grid, z_grid, Q = arrays
+
+    # Obtain ordinary (multi-index) versions of r_σ and P_σ
+    r_σ = compute_r_σ(σ, constants, sizes, arrays)
+    P_σ = compute_P_σ(σ, constants, sizes, arrays)
+
+    # Reshape r_σ and P_σ for a single index state
+    n = y_size * z_size
+    P_σ = jnp.reshape(P_σ, (n, n))
+    r_σ = jnp.reshape(r_σ, n)
+
+    # Solve
+    v_σ = jnp.linalg.solve(jnp.identity(n) - β * P_σ, r_σ)
+
+    # Return as multi-index array
+    return jnp.reshape(v_σ, (y_size, z_size))
+
+get_value_matrix_version = jax.jit(get_value_matrix_version, static_argnums=(2,))
+```
+
++++ {"user_expressions": []}
+
+Similarly, let's define the matrix version to compute $\sigma$-policy operator.
+
+```{code-cell} ipython3
+def T_σ_matrix_version(v, σ, constants, sizes, arrays):
+    "The σ-policy operator, single index version."
+
```

Comment:
@jstac We don't actually use these functions and I guess they were just for verifying. Should we remove them? as we already have other faster function computing the same.